app/src/main/java/streetwalrus/usbmountr/PathResolver.java

```
@@ -22,9 +25,49 @@ public static String getPath(Context context, Uri uri) {
+									// Based on http://stackoverflow.com/questions/28605278/android-5-sd-card-label and https://gist.github.com/prasad321/9852037
+									StorageManager mStorageManager = (StorageManager) context.getSystemService(Context.STORAGE_SERVICE);
+
+									try {
+										Class<?> storageVolumeClazz = Class.forName("android.os.storage.StorageVolume");
+										Method getVolumeList = mStorageManager.getClass().getMethod("getVolumeList");
+										Method getUuid = storageVolumeClazz.getMethod("getUuid");
+										Method getState = storageVolumeClazz.getMethod("getState");
+										Method getPath = storageVolumeClazz.getMethod("getPath");
+										Method isPrimary = storageVolumeClazz.getMethod("isPrimary");
+										Method isEmulated = storageVolumeClazz.getMethod("isEmulated");
+
+										Object result = getVolumeList.invoke(mStorageManager);
+
+										final int length = Array.getLength(result);
+										for (int i = 0; i < length; i++) {
+											Object storageVolumeElement = Array.get(result, i);
+											//String uuid = (String) getUuid.invoke(storageVolumeElement);
+
+											final boolean mounted = Environment.MEDIA_MOUNTED.equals( getState.invoke(storageVolumeElement) )
+													|| Environment.MEDIA_MOUNTED_READ_ONLY.equals(getState.invoke(storageVolumeElement));
+
+											//if the media is not mounted, we need not get the volume details
+											if (!mounted) continue;
+
+											//Primary storage is already handled.
+											if ((Boolean)isPrimary.invoke(storageVolumeElement) && (Boolean)isEmulated.invoke(storageVolumeElement)) continue;
+
+											String uuid = (String) getUuid.invoke(storageVolumeElement);
+
+											if (uuid != null && uuid.equals(type))
+											{
+												String res =getPath.invoke(storageVolumeElement) + "/" +split[1];
+												return res;
+											}
+										}
+									}
+									catch (Exception ex) {
```

This looks fishy.

---

That block can emit ClassNotFoundException, NoSuchMethodException and IllegalAccessException, want me to do something more specific with them? Like show a Toast or something?

---

Failing silently isn't really a good idea, it'd be a good idea to show an error depending on what conditions can lead to this case.

---

Alright, I added a really simple error message for if it fails. It's probably best if we get the full error data if it does though since I can't really predict what situation in particular would be most likely to cause those errors.